include/ppx/application.h

```diff
@@ -526,6 +549,10 @@ class Application
         double   framerateRecordTimer   = 0.0;
         uint64_t framerateFrameCount    = 0;
         bool     resetFramerateTracking = true;
+
+        MetricID                                              nextMetricID = kInvalidMetricID + 1;
+        std::unordered_map<MetricID, metrics::MetricGauge*>   gauges;
```

Comment:
Would it be cleaner if we move more of the record keeping into the metrics Run class and move `MetricID`, `MetricType`, `kInvalidMetricID` into `metrics.h`? Then Run's `AddMetric` can return a MetricID instead of a pointer and provide a separate function to get a metric by ID? It seems like the metrics run would be a more logical place to do the bookkeeping for the IDs since they are tied to the run lifetime anyway. Thoughts?

Reply:
Yes. Using opaque handles solves the issues we have today, and I'm fine converting all APIs to opaque pointers. (IIRC Angela had this pattern the the initial knob API, I had the bad idea to suggest otherwise, sorry..).
But doing it at the application-level seems weird. As Aliya said, moving that at the metrics API level would be better.

Reply:
I'd considered this, and am fine with doing it, BUT it changes the semantics of the metrics API.

Right now, at the metrics level it's possible to have multiple simultaneous runs and different metrics associated with each. The application's specific USE of the metrics API, however, limits things to a single active run. This makes adding metrics trivial: you can call into the application to add a metrics to the active run; if there's no active run, no metric will be added.

Moving it into the metrics API means now the manager needs to take this into account. AddMetric will need to be in the manager to avoid the stale run scenario, which means AddMetric will either need to take a 'runId' argument, or the manager itself will be limited to having a single active run.

I don't personally see a reason for Manager to have more than a single active run, honestly, so I'm totally fine with locking things down at that level, and if we ultimately decide later that's not what we need, we can change things to suit.